components/encryption/src/io.rs

```diff
@@ -644,4 +646,106 @@ mod tests {
             assert_eq!(plaintext, written);
         }
     }
+
+    #[test]
+    fn test_do_crypter_in_place() {
+        let iv = Iv::new_ctr();
+        let key = iv.as_slice().clone();
+        let mut crypter_core =
+            CrypterCore::new(EncryptionMethod::Aes128Ctr, key, Mode::Encrypt, iv).unwrap();
+
+        let mut buf1 = "pingcap 01234567890abcdefg".as_bytes().to_vec();
+        let mut buf2 = buf1.clone();
+        let half_len = buf1.len() / 2;
+
+        let r = crypter_core.do_crypter_in_place(&mut buf1[..]);
+        assert!(r.is_ok());
+
+        let r = crypter_core.reset_crypter(0);
+        assert!(r.is_ok());
+        let r = crypter_core.do_crypter_in_place(&mut buf2[..half_len]);
+        assert!(r.is_ok());
+        let r = crypter_core.do_crypter_in_place(&mut buf2[half_len..]);
+        assert!(r.is_ok());
+        assert_eq!(buf1, buf2);
+    }
+
+    struct MockReader {
```

Comment:
```suggestion
    struct AsyncCursor {
```

Reply:
I think MockReader is more proper, because this reader is a simulation reader for async.

Reply:
Okay, I don't have strong preference against it.

But I just noticed your test didn't really cover the bug. To do that you have to read a partial block: `encrypt_reader.read(&mut encrypt_text[..some_smaller_len])`.

Reply:
Oh, this case should read async multiply, I will modify it ~